### PR TITLE
Use Marker for Unsaved sg.Table Row / Big Performance Improvements / Change attrs["virtual"] to be a list of pks

### PR DESCRIPTION
**Pull request by ssweber**
Do you like this? Marking both unsaved (along with virtual)

https://user-images.githubusercontent.com/57631333/236514752-213d510c-a4d4-47cc-ad1f-a56843efc813.mp4

### CODE REVIEW

**Comment by ssweber** [comment]
This would only apply if either CellEdit / LiveUpdate were enabled, used.

**Comment by ssweber** [comment]
Second commit changes rows.attrs["virtual"] to be a list of pks. I think you'll see it makes reasoning about it much more easy.
Also rewrote table_values and combobox_values to use .apply(). WAY faster. 

I've found asking chatgpt "is there a more performant way of writing this function" will get me alot closer, although I'll end up fixing a few misunderstandings.

With a person table of 15625 records, (1 table, plus 2 combo's being filled with descriptions), a form was basically unusable.

BEFORE:

https://user-images.githubusercontent.com/57631333/236644210-f31a30d4-c5b6-4efb-a4e6-1e91f25d1b2e.mp4

After:

https://user-images.githubusercontent.com/57631333/236644218-c3e5ecbd-a0de-448f-bf7a-608825a0268d.mp4

File: pysimplesql/pysimplesql.py [hunk 1]
```diff
@@ -2527,10 +2519,10 @@ def purge_virtual(self) -> None:
         """
         # remove the rows where virtual is True in place, along with the corresponding
         # virtual attribute
-        idx_to_remove = [idx for idx, v in self.rows.attrs["virtual"].items() if v]
-        self.rows.drop(index=idx_to_remove, inplace=True)
-        for idx in idx_to_remove:
-            del self.rows.attrs["virtual"][idx]
+        virtual_rows = self.rows[self.rows[self.pk_column].isin(self.virtual_pks)]
+        self.rows.drop(index=virtual_rows.index, inplace=True)
+        self.rows.reset_index(drop=True, inplace=True)
```
Comment (ssweber) [inline pysimplesql/pysimplesql.py:2524]: Gonna check in this line…

**Comment by ssweber** [comment]
There’s further room for improvement for updating sg.Tables in LiveUpdate, insert, delete, first, prev, next, last - by directly calling the tkinter Treeview functions so that we arnt re-creating the entire widget each time.

but that’s for another pull :)

**Comment by ssweber** [comment]
Another reason to store pks in rows.attrs[“virtual”]… currently if prompt_save is off, adding multiple virtual rows assigns them all the same pk

This way we can call either call next_pk, or if vrows, increment by max+1.

thinking about this further, will probably change attrs[“row_backup”] to be a dict {pk : backup}, or just a true df. That would get us closer to allowing multiple unsaved changes going— if someone wants to have a sg.Table open almost like excel, and only need to save when quitting. Would need to modify save-records to have another section like we have for keyed-queries that loops through…

**Review by PySimpleSQL — APPROVED** [review]

**Comment by PySimpleSQL** [comment]
Finally have some computer time today :)

**Comment by PySimpleSQL** [comment]
These sound good to me.

I am finally at a computer again.  I had to redo my travel computer due to
the proxy stuff being a real pain in Linux (bash had its own proxy
settings, Git had its own, the system had proxy settings, firefox had its
own, PyCharm had separate proxy settings.... So I kept having to redo these
depending on whether I had to use PDANet+ on my phone or not..  I finally
have USB tethering set up so no more proxy madness, and I can get back to
work on the project

**Comment by ssweber** [comment]
Yeah that sounds complicated. With all your traveling sounds like you need one of those wifi hotspot jet packs.

**Comment by PySimpleSQL** [comment]
We have starlink, I should just mount it on the RV for our trips.

I had unlimited hotspot data and somehow got duped out of the plan when I
upgraded my phone, so I think I finally got around the issue with usb
tethering.  Hopefully anyways!

**Pull request merged**